dask/array/gufunc.py

```diff
@@ -0,0 +1,455 @@
+    if isinstance(signature, str):
+        core_input_dimss, core_output_dimss = _parse_gufunc_signature(signature)
+    else:
+        core_input_dimss, core_output_dimss = signature
+
+    ## Determine nout
+    nout = None if not isinstance(core_output_dimss, list) else len(core_output_dimss)
+
+    ## Assert output_dtypes
+    if nout is not None and not (isinstance(output_dtypes, tuple) or isinstance(output_dtypes, list)):
+        raise ValueError("Must specify tuple of dtypes for `output_dtypes` for function with multiple outputs")
+    if nout is None and (isinstance(output_dtypes, tuple) or isinstance(output_dtypes, list)):
+        raise ValueError("Must specify single dtype for `output_dtypes` for function with one output")
+
+    ## Use top to apply func
+    if vectorize:
+        signature = _compile_gufunc_signature(core_input_dimss, core_output_dimss)
+        func = np.vectorize(func, signature=signature)
+
+    ## Miscellaneous
+    if output_sizes is None:
+        output_sizes = {}
+
+    # Main code:
+    ## Cast all input arrays to dask
+    args = [asarray(a) for a in args]
+
+    if len(core_input_dimss) != len(args):
+        ValueError("According to `signature`, `func` requires %d arguments, but %s given"
+                   % (len(core_output_dimss), len(args)))
+
+    ## Assess input args for loop dims
+    input_shapes = [a.shape for a in args]
+    input_chunkss = [tuple(c[0] for c in a.chunks) for a in args]
+    num_loopdims = [len(s) - len(cd) for s, cd in zip(input_shapes, core_input_dimss)]
+    max_loopdims = max(num_loopdims) if num_loopdims else None
+    _core_input_shapes = [dict(zip(cid, s[n:])) for s, n, cid in zip(input_shapes, num_loopdims, core_input_dimss)]
+    core_shapes = merge(output_sizes, *_core_input_shapes)
+
+    loop_input_dimss = [tuple("__loopdim%d__" % d for d in range(max_loopdims - n, max_loopdims)) for n in num_loopdims]
+    input_dimss = [l + c for l, c in zip(loop_input_dimss, core_input_dimss)]
+
+    loop_output_dims = max(loop_input_dimss, key=len) if loop_input_dimss else set()
+
+    ## Assess input args for same size and chunk sizes
+    ### Check that the arrays have same length for same dimensions or dimension `1`
+    _temp = groupby(0, concat(zip(ad, s) for ad, s in zip(input_dimss, input_shapes)))
+    dimsizess = valmap(compose(set, partial(map, itemgetter(1))), _temp)
+    if not allow_rechunk:
+        for dim, sizes in dimsizess.items():
+            if sizes.union({1}) != {1, max(sizes)}:
+                raise ValueError("Dimension ``{}`` with different lengths in arrays".format(dim))
+
+    # Check if arrays have same chunk size for the same dimension
+    _temp = groupby(0, concat(zip(ad, s, c) for ad, s, c in zip(input_dimss, input_shapes, input_chunkss)))
+    dimchunksizess = valmap(compose(tuple, unique,
+                                    partial(map, itemgetter(1)),
+                                    partial(filter, lambda e: e != (1, 1)),
```

Comment:
What does`(1, 1)` indicate here?

(For what it's worth, I find these sort of expressions much clearer when written imperatively in standard Python style with good variable names rather than with toolz.)

Reply:
Checking consistency of chunk sizes requires to consider the respective dimsize as well. `(1, 1)` is a combination of `(<dimsize>, <chunksize>)` which is OK already. Let me find a better implementation of these lines of code

Note for myself:
[ ] Reimplement a more readble consistency validation of chunk sizes as proposed

Reply:
Hopefully the new implementation is more readable